File: tests/data-conversion/union/union.cpp
#include "runtime/union.h"

namespace {

void TakeOptionalUnion(compilets::Union<std::monostate, double, bool> a) {
  if (!compilets::IsTrue(a)) {}
}

void TakeNumber(double n) {}

compilets::Union<double, bool> ReturnUnion() {
  return static_cast<double>(123);
}

void TestUnion() {
  compilets::Union<double, bool> bn = static_cast<double>(999);
  bn = true;
  TakeOptionalUnion(bn);
  TakeOptionalUnion(static_cast<double>(888));
  TakeOptionalUnion(true);
  compilets::Union<double, bool> nb = ReturnUnion();
  bn = ReturnUnion();
  bn = nb;
  bool b = std::get<bool>(bn);
  TakeNumber(std::get<double>(bn));
  double numberCast = std::get<double>(bn);
}

}  // namespace

File: tests/data-conversion/union/union.ts
function TakeOptionalUnion(a?: boolean | number) {
  if (!a) {
  }
}

function TakeNumber(n: number) {
}

function ReturnUnion(): boolean | number {
  return 123;
}

function TestUnion() {
  let bn: boolean | number = 999;
  bn = true;
  TakeOptionalUnion(bn);
  TakeOptionalUnion(888);
  TakeOptionalUnion(true);

  let nb: number | boolean = ReturnUnion();
  bn = ReturnUnion();
  bn = nb;

  let b = bn as boolean;
  TakeNumber(bn as number);

  let numberCast = bn as number;
}
